### ISSUE TITLE
OTP Authentication Malfunctioning

### ISSUE THREAD

**Issue by allahshukur-ahmadzada** [PowerDNS-Admin/PowerDNS-Admin#1403]
Hi, I am trying to setup OTP. from basic settings forced OTP and when I login it shows QR code I scan with Google auth and it gets registered, but after submitting the code I get 500 error

```
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1519, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1517, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1503, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/app/powerdnsadmin/routes/index.py", line 737, in welcome
    good_token = user.verify_totp(otp_token)
  File "/app/powerdnsadmin/models/user.py", line 98, in verify_totp
    return totp.verify(token, valid_window = 5)
  File "/usr/lib/python3.10/site-packages/pyotp/totp.py", line 80, in verify
    if utils.strings_equal(str(otp), str(self.at(for_time, i))):
  File "/usr/lib/python3.10/site-packages/pyotp/totp.py", line 56, in at
    return self.generate_otp(self.timecode(for_time) + counter_offset)
  File "/usr/lib/python3.10/site-packages/pyotp/otp.py", line 35, in generate_otp
    hasher = hmac.new(self.byte_secret(), self.int_to_bytestring(input), self.digest)
  File "/usr/lib/python3.10/site-packages/pyotp/otp.py", line 49, in byte_secret
    missing_padding = len(secret) % 8

```

**Comment by allahshukur-ahmadzada** [comment]
@AzorianMatt  currently there is no way to make OTP work?  I see last release was on summer since that time no one noticed issue? Do you think product okay to be used for production environments?

**Comment by AzorianMatt** [comment]
@allahshukur-ahmadzada as far as I now, OTP works just fine. I'm not sure what issue you're having here exactly, but we'll hopefully get to figuring it out. One of the other contributors was just doing testing last week involving OTP so it presumably works. The product is okay for production environments and is currently being used by many as such. If you're having issues with the main branch of the project which is a bit unstable right now, I recommend you switch to the 0.3.0 release that is stable.

**Comment by allahshukur-ahmadzada** [comment]
Thank you @AzorianMatt, that is how I guessed , I did like this from settings makde otp_force enabled and on next login it showed me QR for OTP registration, now sure if this is right way.

**Comment by AzorianMatt** [comment]
@allahshukur-ahmadzada you're welcome. Admittedly, I have not really played around with the global OTP setting thus far. I believe if you just enable it on a per user basis, that much is at least properly functional. I will try to make it a point to test this further in the coming week or so as time permits.

**Comment by subbink** [comment]
I cannot reproduce the error you mention when I use the current dev branch.
I changed settings > Basic > otp_force to On (this was Off) and otp_field_enabled was already On.
Then I logged out and logged in again. This presented me with a screen to setup OTP. With that information I entered a OTP code on the screen and then on the login screen I was able to login using my username, password and OTP code.

**Comment by AzorianMatt** [comment]
@allahshukur-ahmadzada will you please verify if you still have this issue using the latest release of 0.4.0?

**Comment by AzorianMatt** [comment]
@allahshukur-ahmadzada I'm closing this issue as I was unable to reproduce it using the current `dev` branch which is in step with the `0.4.1` release. As far as it relates to this code, `0.4.0` should also prove to work as intended.

**Comment by Zebble** [comment]
I'm having this same issue.

I'm using LDAP (AD) as the external authentication provider.  When I force OTP, I get the OTP setup screen with the QR code.  When I scan it and enter the code, I get a 500 error and the tracebacks below.

One note that might help is that when the OTP setup screen pops up, "None" is displayed in the "Your secret key is:" box, instead of the key.  There is no debug error in the logs prior or after the OTP setup screen is shown, only after the code is first entered.

Tracebacks:

Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1519, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1517, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1503, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/app/powerdnsadmin/routes/index.py", line 864, in welcome
    good_token = user.verify_totp(otp_token)
  File "/app/powerdnsadmin/models/user.py", line 99, in verify_totp
    return totp.verify(token, valid_window = 5)
  File "/usr/lib/python3.10/site-packages/pyotp/totp.py", line 80, in verify
    if utils.strings_equal(str(otp), str(self.at(for_time, i))):
  File "/usr/lib/python3.10/site-packages/pyotp/totp.py", line 56, in at
    return self.generate_otp(self.timecode(for_time) + counter_offset)
  File "/usr/lib/python3.10/site-packages/pyotp/otp.py", line 35, in generate_otp
    hasher = hmac.new(self.byte_secret(), self.int_to_bytestring(input), self.digest)
  File "/usr/lib/python3.10/site-packages/pyotp/otp.py", line 49, in byte_secret
    missing_padding = len(secret) % 8
TypeError: object of type 'NoneType' has no len()
[2023-04-27 13:12:56,788] [app.py:1449] ERROR - Exception on /welcome [POST]
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1519, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1517, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1503, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/app/powerdnsadmin/routes/index.py", line 864, in welcome
    good_token = user.verify_totp(otp_token)
  File "/app/powerdnsadmin/models/user.py", line 99, in verify_totp
    return totp.verify(token, valid_window = 5)
  File "/usr/lib/python3.10/site-packages/pyotp/totp.py", line 80, in verify
    if utils.strings_equal(str(otp), str(self.at(for_time, i))):
  File "/usr/lib/python3.10/site-packages/pyotp/totp.py", line 56, in at
    return self.generate_otp(self.timecode(for_time) + counter_offset)
  File "/usr/lib/python3.10/site-packages/pyotp/otp.py", line 35, in generate_otp
    hasher = hmac.new(self.byte_secret(), self.int_to_bytestring(input), self.digest)
  File "/usr/lib/python3.10/site-packages/pyotp/otp.py", line 49, in byte_secret
    missing_padding = len(secret) % 8
TypeError: object of type 'NoneType' has no len()

**Comment by AzorianMatt** [comment]
@Zebble that note about the nun value in the secret key is valuable to know. It seems that would be a good indicator to follow for finding the cause of this issue.

Thank you for your addition.

**Comment by Zebble** [comment]
Can this be-reopened then or should I open a new issue?
